cmd/zfs/zfs_main.c

```diff
@@ -7195,14 +7212,23 @@ share_mount(int op, int argc, char **argv)
 			argv++;
 		}
 
-		if (argc != 0) {
-			(void) fprintf(stderr, gettext("too many arguments\n"));
+		if (argc > 1) {
+			(void) fprintf(stderr, gettext("usage: "
+			    "zfs mount -a <dataset>\n"));
 			usage(B_FALSE);
 		}
 
+		if (recursive && argc != 1) {
+			(void) fprintf(stderr, gettext("usage: "
+			    "zfs mount -r <dataset>\n"));
+			usage(B_FALSE);
+		}
+
+		const char *dataset = (argv[0] != NULL) ? argv[0] : NULL;
```

Comment:
@ahesford raised 3 points regarding dataset

that the line should read const char *dataset = arvg[0]
we need to add a check to ensure dataset, is in fact actually a valid (existing) dataset
to improve clarity all instances dataset should be changed to filesystem to be inline with documentation